Java/parkservices/src/main/java/com/amazon/randomcutforest/parkservices/GlobalLocalAnomalyDetector.java

```diff
@@ -0,0 +1,410 @@
+import com.amazon.randomcutforest.summarization.ICluster;
+import com.amazon.randomcutforest.summarization.Summarizer;
+import com.amazon.randomcutforest.util.Weighted;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.function.BiFunction;
+
+import static com.amazon.randomcutforest.CommonUtils.checkArgument;
+import static com.amazon.randomcutforest.summarization.GenericMultiCenter.DEFAULT_NUMBER_OF_REPRESENTATIVES;
+import static com.amazon.randomcutforest.summarization.GenericMultiCenter.DEFAULT_SHRINKAGE;
+import static java.lang.Math.abs;
+import static java.lang.Math.min;
+
+public class GlobalLocalAnomalyDetector<P> extends StreamSampler<P> {
+
+    // default maximum number of clusters to consider
+    public static int DEFAULT_MAX = 10;
+
+    // an upper bound on the score
+    public static float FLOAT_MAX = 10;
+
+    // the relative weight of small clusters which should not be used in anomaly
+    // detection
+    // this controls masking effects
+    public static double DEFAULT_IGNORE_SMALL_CLUSTER_REPRESENTATIVE = 0.005;
+
+    // the number of steps we have to wait before reclustering; in principle this
+    // can be 1, but that would be
+    // neither be meaningful nor efficient; it is set to a default of the capacity/2
+    protected int doNotreclusterWithin;
+
+    // a thresholder for flagging anomalies (same thresholder as in TRCF)
+    protected final BasicThresholder thresholder;
+
+    // remembering when the last clustering was performed
+    protected long lastCluster = 0L;
+
+    // remembers when the mean of the scores were just above a certain threshold
+    // acts as a calibration mechanism
+    protected double lastMean = 1;
+
+    // actual list of clusters
+    List<ICluster<P>> clusters;
+
+    // the number of maximum clusters to be considered; this is configurable and can
+    // be chaned dynamically
+    protected int maxAllowed;
+
+    // the shrinkage parameter in multi-centroid clustering such as CURE. Shrinkage
+    // of 0 provides
+    // non-spherical shapes, whereas shrinkage of 1 corresponds to choosing single
+    // centroid (not recommended)
+    protected double shrinkage;
+
+    // number of representatives used in multi-centroidal clustering
+    protected int numberOfRepresentatives;
+
+    // threshold of weight for small clusters so that masking can be averted, can be
+    // changed dynamically
+    protected double ignoreBelow;
+
+    // the global function used in clustering, can be changed dynamically (but
+    // clustering would be controlled
+    // automatically due to efficiency reasons)
+    protected BiFunction<P, P, Double> globalDistance;
+
+    public static Builder<?> builder() {
+        return new Builder<>();
+    }
+
+    protected GlobalLocalAnomalyDetector(Builder<?> builder) {
+        super(builder);
+        thresholder = new BasicThresholder(builder.timeDecay);
```

Comment:
There is no setting currently such as expected anomaly rate -- primarily because TRCF dealt with metrics where there is continuity in time; whereas here each object stands on its own. That being said; this is a good callout -- perhaps it would be useful to define anomalyRate to control burstiness as well as for future/potential use. Right now it is the same as the time decay of the sampler.